nemo_aligner/utils/batch_iterators.py

```diff
@@ -0,0 +1,151 @@
+
+    def clear(self):
+        with self.lock:
+            self.data.clear()
+
+    def set_idx(self, ids):
+        with self.lock:
+            self.data.update(ids)
+
+    def get_idx(self, batch_size):
+        with self.lock:
+            to_ret = [self.data.pop() for _ in range(batch_size) if len(self.data) > 0]
+
+        return to_ret
+
+
+@dataclass
+class DefaultBatchIterator:
+    """The default batch iterator used for getting samples for generation stage 
+    """
+
+    sampler_iter: Iterator[int]
+    num_microbatches: int
+    dataset: Mapping
+    collate_fn: Callable
+
+    def __iter__(self):
+        for _, ids in zip(range(self.num_microbatches), self.sampler_iter):
+            batch = self.collate_fn([self.dataset[index] for index in ids])
+            yield batch
+
+
+@dataclass
+class HTTPBatchIterator:
+    shared_set: Union[SharedSet, None]
+    host: str
+    port: int
+    sampler_iter: Iterator[int]
+    num_microbatches: int
+    dataset: Mapping
+    collate_fn: Callable
+
+    def __post_init__(self):
+        local_ids = [ids for _, ids in zip(range(self.num_microbatches), self.sampler_iter)]
+        self.desired_batch_size = len(local_ids[0]) if len(local_ids) > 0 else 1
+
+        local_ids = set(itertools.chain.from_iterable(local_ids))
+        global_ids = get_global_set(local_ids)
+
+        if torch.distributed.get_rank() == 0:
+            self.shared_set.clear()
+            self.shared_set.set_idx(global_ids)
+
+        torch.distributed.barrier()
+
+    def __iter__(self):
+        ids = _send_request(host=self.host, port=self.port, batch_size=self.desired_batch_size)
+
+        while len(ids) > 0:
+            batch = self.collate_fn([self.dataset[idx] for idx in ids])
+            yield batch
+            ids = _send_request(host=self.host, port=self.port, batch_size=self.desired_batch_size)
+
+        if torch.distributed.get_rank() == 0:
+            self.shared_set.clear()
+
+
+def get_batch_iterator_cls(batch_iterator_cfg):
+    use_flask = batch_iterator_cfg.get("use_flask", False)
+
+    if use_flask:
+        port = batch_iterator_cfg.get("port", 1234)
+        ip_address = [socket.gethostbyname(socket.gethostname())]
+        torch.distributed.broadcast_object_list(ip_address, src=0, group=None, device=torch.cuda.current_device())
+        flask_host = ip_address[0]
+        shared_set = None
+
+        # start the server on rank 0
+        if torch.distributed.get_rank() == 0:
+            lock = threading.Lock()
+
+            shared_set = SharedSet(lock)
+            app = Flask(__name__)
+
+            @app.route("/get_idx", methods=["PUT"])
+            def get_http_idx():
+                batch_size = request.get_json()["batch_size"]
+                return shared_set.get_idx(batch_size)
+
+            flask_thread = threading.Thread(
+                target=lambda: app.run(host=flask_host, port=port, use_reloader=False), daemon=True
+            )
+            flask_thread.start()
```

Comment:
because it's a daemon thread I don't think we need to

from [python doc](https://docs.python.org/3/library/threading.html#thread-objects)
```
A thread can be flagged as a “daemon thread”. The significance of this flag is that the entire Python program exits when only daemon threads are left. The initial value is inherited from the creating thread. The flag can be set through the [daemon](dfile:///Users/geshen/Library/Application%20Support/Dash/DocSets/Python_3/Python%203.docset/Contents/Resources/Documents/doc/library/threading.html#threading.Thread.daemon) property or the daemon constructor argument.
```

Reply:
Ah I see -- but below that it also says:
"""
Daemon threads are abruptly stopped at shutdown. Their resources (such as open files, database transactions, etc.) may not be released properly. If you want your threads to stop gracefully, make them non-daemonic and use a suitable signalling mechanism such as an [Event](https://docs.python.org/3/library/threading.html#threading.Event).
"""
=> it's not obvious that it's going to cleanup nicely.

I guess it's probably not a big deal though -- especially since we're working in container images -- but maybe you can just add a TODO comment like
```
# TODO Starting the thread as daemon means it may not release all resources on exit => maybe implement a clean shutdown logic?
```

Reply:
oh no i completely missed that... added 4bdf27be58305bfb1c6cc3ed93848e5daaf203b8